File: app/src/main/java/com/imdglobalservices/psi/network/api/RestClientService.java
package com.imdglobalservices.psi.network.api;

import com.imdglobalservices.psi.BuildConfig;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author Dwi Setiyono <dwi.setiyono@ovo.id>
 * @since 2017.15.11
 */
public class RestClientService {
    private static RestClientService instance;
    private static HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();

    private RestClientService() {
    }

    public static synchronized RestClientService getInstance() {
        if (instance == null) {
            instance = new RestClientService();
            loggingInterceptor.setLevel(getEnvironmentLevel(BuildConfig.ENVIRONMENT));
        }

        return instance;
    }

    private static HttpLoggingInterceptor.Level getEnvironmentLevel(int env) {
        return env == 0 ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.NONE;
    }

    /**
     * this method rest client will running without retry function
     *
     * @return RestClient
     */
    public RestClient getConnections() {
        return runningClient(null, false);
    }

    /**
     * this method rest client will running without retry function
     *
     * @return RestClient
     */
    public RestClient getConnectionsWithRetry() {
        return runningClient(null, true);
    }

    private RestClient runningClient(RestClient restClient, boolean isRetry) {
        int TIMEOUT = 180000;

        if (restClient == null) {
            OkHttpClient client = new OkHttpClient.Builder()
                    .addInterceptor(loggingInterceptor)
                    .retryOnConnectionFailure(isRetry)
                    .connectTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
                    .readTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
                    .writeTimeout(TIMEOUT, TimeUnit.MILLISECONDS)
                    .build();
            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl(APIDomain.getApiBaseURL())
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
            restClient = retrofit.create(RestClient.class);
        }
        return restClient;
    }
}

File: app/src/main/java/com/imdglobalservices/psi/network/api/RestClientService.kt
package com.imdglobalservices.psi.network.api

import com.imdglobalservices.psi.BuildConfig
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * @author Dwi Setiyono <dwi.setiyono></dwi.setiyono>@ovo.id>
 * @since 2017.15.11
 */
class RestClientService private constructor() {

    /**
     * this method rest client will running without retry function
     *
     * @return RestClient
     */
    val connections: RestClient?
        get() = runningClient(null, false)

    /**
     * this method rest client will running without retry function
     *
     * @return RestClient
     */
    val connectionsWithRetry: RestClient?
        get() = runningClient(null, true)

    private fun runningClient(restClient: RestClient?, isRetry: Boolean): RestClient? {
        var restClient = restClient
        val TIMEOUT = 180000

        when (restClient) {
            null -> {
                val client = OkHttpClient.Builder()
                        .addInterceptor(loggingInterceptor)
                        .retryOnConnectionFailure(isRetry)
                        .connectTimeout(TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                        .readTimeout(TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                        .writeTimeout(TIMEOUT.toLong(), TimeUnit.MILLISECONDS)
                        .build()
                val retrofit = Retrofit.Builder()
                        .baseUrl(APIDomain.apiBaseURL)
                        .client(client)
                        .addConverterFactory(GsonConverterFactory.create())
                        .build()
                restClient = retrofit.create(RestClient::class.java)
            }
        }
        return restClient
    }

    companion object {
        internal var instance: RestClientService? = null
        private val loggingInterceptor = HttpLoggingInterceptor()

        @Synchronized
        fun getInstance(): RestClientService? {
            if (instance == null) {
                instance = RestClientService()
                loggingInterceptor.level = getEnvironmentLevel(BuildConfig.ENVIRONMENT)
            }

            return instance
        }

        private fun getEnvironmentLevel(env: Int): HttpLoggingInterceptor.Level {
            return if (env == 0) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE
        }
    }
}
